Refactor preprocessing files.

MarkAZhang opened the pull request:
# Description

- Removed the `reshape_mosquito_per_image` file, which had many common methods with `preprocess_mosquito_image` and added the `allocate_one_mosquito` functionality to `preprocess_mosquito_image`. When the user provides the optional argument `individual_mosquito_dir`, all individual mosquito images will be placed in that directory.
- Moved all preprocessing methods into a separate file `preprocessing.py`.
- Added an additional command `debug_preprocessing_single_image`, which runs the preprocessing on a single image and outputs all images to an `output_dir`. This function also outputs the intermediate `binary_by_saturation_filter` image, which will help us debug any changes we make there.
- Added a `main.py` function that allows us to call the functions from Python, instead of via command line. (I got tired of trying to type these super long commands in the terminal)
- Generalized the `allocate_one_mosquito` image a little bit (renamed it to `get_subimage`)
- Also fixed a couple of issues that were preventing me from running the code, including changing the version of Pandas and changing the function signature on `_, contours, _ = cv2.findContours`. My version of `findContours` only outputs 2 parameters.

pranathivemuri commented:
I like having command lines so people can access and run the code themselves without having to import functions. Its just easier to give them a command and ask them to copy paste instead of giving them a program. I had to just rename the folder names and copy paste the commands for different folders. So, I didn’t type it multiple times

For opencv and pandas - i am not going to change the versions. Please create a conda environment woth environment.yml file requirements

I am going to merge the rest of the parts of your changes that are good refactor requests and merge them into my branch

Thanks!

MarkAZhang commented:
- It's still possible to use the command line directly. I just added an additional way to call it from code. I can rename main.py to test.py so it's clear that it's not the intended way to run the library (and we can remove that file before we publish the library). I find the code interface a little easier to use personally. Let me know if that works.

- I'm a little confused by the `environment.yml` versions comment. I feel like `requirements.txt` should have versions that match what's in the code. I changed the versions because when I tried to run `pip install -r requirements.txt` and then run the Python code I got some errors. Perhaps you can explain a little more.

pranathivemuri commented on debug_preprocessing_single_image.py:
```diff
@@ -0,0 +1,88 @@
+    binary_mosquito_image = binarize_by_saturation_filter(
+        color_mosquito_image, args.min_saturation, args.max_saturation)
+
+    cv2.imwrite(
```
binary image should be saved as x, y shaped image with no rgb channels

pranathivemuri commented on example.py:
```diff
@@ -0,0 +1,50 @@
+
+# Example of calling the library directly from Python. Feel free to modify according to your directory structure.
+# You can also call the library directly from the command line.
+def process_all_mosquiti_images():
```
mosquito spelling

pranathivemuri commented on requirements.txt:
```diff
@@ -8,7 +8,6 @@ click<7,>=6.7
 matplotlib==3.0.3
 itkwidgets==0.14.0
-pandas<0.21
+pandas==0.25.2
```
I changed these lines now in my branch

pranathivemuri replied:
also changed opencv version.

pranathivemuri commented on example.py:
```diff
@@ -0,0 +1,50 @@
+PREPROCESS_MOSQUITO_IMAGES_ARGPARSER = get_preprocess_mosquito_images_argparser()
+DEBUG_PREPROCESSING_SINGLE_IMAGE_ARGPARSER = get_debug_preprocessing_single_image_argparser()
+
+# Example of calling the library directly from Python. Feel free to modify according to your directory structure.
```
sorry I missed this! thanks for the comments and example

pranathivemuri commented on preprocessing.py:
```diff
@@ -208,102 +151,50 @@ def draw_bb_labels_around_contours(
+def get_subimage(
+        color_mosquito_image, bounding_box):
+    """
+    Returns a list of single mosquito images of TILE_SIZE
```
change docstring

pranathivemuri commented on preprocessing.py:
```diff
@@ -208,102 +151,50 @@ def draw_bb_labels_around_contours(
-        image = cv2.imread(path)
-        color_mosquito_image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
+def extract_subimages_by_bounding_boxes(image, bounding_boxes):
+    """
```
correct the docstring to be like other functions descriptions with input args and return args.

pranathivemuri commented:
@MarkAZhang thanks for the refactor, I have a few comments regarding docstrings that didn't look right to me. fix them and we will merge this onto the other PR.
